wgengine/tsdns/tsdns.go

```diff
@@ -0,0 +1,262 @@
+	// For example, monitoring.ipn.dev -> 100.64.0.1.
+	DomainToIP map[string]netaddr.IP
+}
+
+// Resolver is a DNS resolver for domain names of the form *.ipn.dev
+type Resolver struct {
+	logf logger.Logf
+
+	// ip is the IP on which the resolver is listening.
+	ip packet.IP
+	// port is the port on which the resolver is listening.
+	port uint16
+
+	// parser is a request parser that is reused by the resolver.
+	parser dns.Parser
+	// responseBuffer is a static buffer to avoid graticious allocations.
+	responseBuffer [maxResponseSize]byte
+
+	// mu guards the following fields from being updated while used.
+	mu sync.Mutex
+	// dnsMap is the map most recently received from the control server.
+	dnsMap *Map
+}
+
+// NewResolver constructs a resolver with default parameters.
+func NewResolver(logf logger.Logf) *Resolver {
+	return &Resolver{
+		logf: logf,
+		ip:   DefaultIP,
+		port: DefaultPort,
+	}
+}
+
+// AcceptsPacket determines if the given packet is
+// directed to this resolver (by ip and port).
+// We also require that UDP be used to simplify things for now.
+func (r *Resolver) AcceptsPacket(in *packet.ParsedPacket) bool {
+	return in.DstIP == r.ip && in.DstPort == r.port && in.IPProto == packet.UDP
+}
+
+// SetMap sets the resolver's DNS map.
+func (r *Resolver) SetMap(m *Map) {
+	r.mu.Lock()
+	r.dnsMap = m
+	r.mu.Unlock()
+}
+
+// Resolve maps a given domain name to the IP address of the host that owns it.
+func (r *Resolver) Resolve(domain string) (netaddr.IP, dns.RCode, error) {
+	// If not a subdomain of ipn.dev, then we must refuse this query.
+	// We do this before checking the map to distinguish beween nonexistent domains
+	// and misdirected queries.
+	if !strings.HasSuffix(domain, ".ipv.dev") {
+		return netaddr.IP{}, dns.RCodeRefused, errNotOurName
+	}
+
+	r.mu.Lock()
+	if r.dnsMap == nil {
+		r.mu.Unlock()
+		return netaddr.IP{}, dns.RCodeServerFailure, errMapNotSet
+	}
+	addr, found := r.dnsMap.DomainToIP[domain]
+	r.mu.Unlock()
+
+	if !found {
+		return netaddr.IP{}, dns.RCodeNameError, errNoSuchDomain
+	}
+	return addr, dns.RCodeSuccess, nil
+}
+
+type response struct {
+	Header         dns.Header
+	ResourceHeader dns.ResourceHeader
+	Question       dns.Question
+	IP             netaddr.IP
+}
+
+func (r *Resolver) parseQuery(query *packet.ParsedPacket, resp *response) error {
+	var err error
+
+	resp.Header, err = r.parser.Start(query.Payload())
+	if err != nil {
+		resp.Header.RCode = dns.RCodeFormatError
+		return err
+	}
+
+	if resp.Header.Response {
+		resp.Header.RCode = dns.RCodeFormatError
+		return errNotQuery
+	}
+
+	resp.Question, err = r.parser.Question()
+	if err != nil {
+		resp.Header.RCode = dns.RCodeFormatError
+		return err
+	}
+
+	return nil
+}
+
+func (r *Resolver) makeResponse(resp *response) error {
+	var err error
+
+	switch resp.Question.Type {
+	case dns.TypeA, dns.TypeALL:
+		// Remove final dot from name: *.ipn.dev. -> *.ipn.dev
+		name := resp.Question.Name.String()
+		name = name[:len(name)-1]
```

Comment:
Guaranteed to end in a "." and to be non-empty? Otherwise this will panic or corrupt the response.